### PR TITLE
Filter records

### LINKED ISSUE #25: Filter records
We need to be able to filter records from `impagination`. Otherwise, when we try and filter from a parent interface, we lose the ability to lazily get records.

### LINKED ISSUE #27: Apply actions on the dataset
- [x] applyFilter - reapply all filters to the resolved pages (handled by refresh)
- [x] refresh - refreshes the dataset at the present/current read offset
- [x] reset - resets the dataset to the readOffset 0

### PR DESCRIPTION
**Pull request by flexyford**
Fix #25 Filtering records on Resolved Pages is applied whenever a fetch returns. 

Previously each page had `records`, but this has been extended to `unfilteredRecords` and `records` (where `records` are now records which include all records which passed through the filter).

We also introduce a `filter` function on the dataset, which applies the filter every time a page is resolved!

Fix #27 
Introduce actions for refreshing the dataset (aka reapplying filters) and resetting the dataset (unloading all pages and refetch the dataset).
refresh(readOffset) will reapply filters on all pages from the given offset (defaultOffset: currentOffset)
reset(readOffset) will unload and refetch all pages from the given offset (defaultOffset: 0)

### CODE REVIEW

**Comment by flexyford** [comment]
Woops, checked0in way too many files :-1:

File: src/dataset.js [hunk 1]
```diff
@@ -182,7 +203,8 @@ export default class Dataset {
       let state = this.state.update((next)=> {
         next.stats = stats;
         if(page !== next.pages[offset]) { return; }
-        next.pages[offset] = page.resolve(records);
+        // Filter on page update
+        next.pages[offset] = page.resolve(records).filter(this._filter);
```
Comment (flexyford) [inline src/dataset.js]: Should the filter be applied automatically when a page resolves? That way, when we want to reapply the filter to a resolved page, we just call `page.resolve` and the new filter is automatically reapplied!

File: src/dataset.js [hunk 2]
```diff
@@ -33,16 +33,36 @@ class State {
...
     let recordOffset = index % this.pageSize;
+
+    // Dynamically find the page offset
+    const minUnloadPage = Math.floor((this.readOffset - this.unloadHorizon) / this.pageSize);
+    const minUnloadHorizon = Math.max(minUnloadPage, 0);
+    // const maxUnloadPage = Math.ceil((this.readOffset  + this.unloadHorizon) / this.pageSize);
+    // const maxUnloadHorizon = Math.min(this.stats.totalPages || Infinity, maxUnloadPage, this.pages.length);
+    if(pageOffset >= minUnloadHorizon) {
+      let _this = {index: index};
+      pageOffset = this.pages.findIndex(function(page) {
```
Comment (cowboyd) [inline src/dataset.js]: We should be able to execute this find inside the slice of pages between `minUnloadPage` and `minUnloadHorizon`

File: src/dataset.js [hunk 3]
```diff
@@ -69,13 +100,21 @@ export default class Dataset {
     }
   }
 
-  setReadOffset(readOffset) {
-    if (this.state.readOffset === readOffset) { return; }
+
+  setReadOffset(readOffset, options = {}) {
```
Comment (cowboyd) [inline src/dataset.js]: The `setReadOffset` function seems too overloaded here. Is there any reason to not have `setReadOffset`, `refresh`, and `refilter` just do only those things?

Comment (flexyford) [inline src/dataset.js]: I'll work to organize this better, but there would be SO MUCH duplicate code, that it made more sense to do setReadOffset with options! Over 96% of the code would be duplicate code.

Comment (flexyford) [inline src/dataset.js]: Latest commit attempts to organize the extra logic more effectivly

File: src/dataset.js [hunk 4]
```diff
@@ -69,13 +100,18 @@ export default class Dataset {
     }
   }
 
-  setReadOffset(readOffset) {
-    if (this.state.readOffset === readOffset) { return; }
+  setReadOffset(readOffset, options = {}) {
+    this._validateOptions(options);
+    this._initPageByOption = this._initPage(options);
```
Comment (flexyford) [inline src/dataset.js]: This attempts to make the options logic more compact within setReadOffset

**Comment by cowboyd** [comment]
I'm still uncomfortable with the intertwining of `setReadOffset`, `refilter` and `reload` actions, and the parameterization of `setReadOffset`.

I'm worried that we're using the [wrong abstraction][1] in order to avoid duplication. In my mind, the refiltering should do nothing except walk the resolved pages, and replace them with a filtered version of the pages. Same thing with reload. 

If we do find out which operations are common between them, then perhaps we can introduce a _new_ abstraction instead of try to piggy back on the existing one.

[1]: http://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction

File: src/dataset.js [hunk 5]
```diff
@@ -121,6 +156,107 @@ export default class Dataset {
...
+
+      // Update the total number of records after applying filters
+      this._adjustTotalRecords(next);
+    });
+    this._observe(this.state = state);
+  }
+
+  // Unload all pages, and refetch from the readOffset (default: 0)
+  // Reload does `unfetch` all unloaded pages
+  reload(readOffset){
```
Comment (cowboyd) [inline src/dataset.js]: What's the use-case for reloading at a different readOffset than the current read offset?

Comment (flexyford) [inline src/dataset.js]: I agree that the default should be the current readOffset.

However, let's say we want to reload our dataset from the beginning. Let's say the current readOffset is 100, and we wish to set it to 0. Reload will then `unfetch` all pages and start at the beginning.

Comment (cowboyd) [inline src/dataset.js]: Ok, I think I'm still trying to understand the use-case difference between reset and reload, vs. re-filter.

Comment (flexyford) [inline src/dataset.js]: I'll write up the README today, and then we can discuss the best course forward. Getting back to it today after stepping away for a month, I was confused too. Documentation will help with the understanding.

**Pull request merged**